admin/modules/config/plugins.php

```diff
@@ -517,36 +680,53 @@
 
 	$plugins->run_hooks("admin_config_plugins_plugin_list");
 
+	$form = new Form('index.php?module=config-plugins&amp;action=upload', 'post', '', 1);
+	$table = new Table;
+	$table->construct_header($lang->upload_plugin_desc);
+	$table->construct_cell($form->generate_file_upload_box('plugin_upload', array('style' => 'width: 230px;')).
+	                       $form->generate_check_box('ignore_vers', 1, $lang->plugin_ignore_vers).
+	                       ' '.
+	                       $form->generate_submit_button($lang->install_uploaded_plugin));
+	$table->construct_row();
+	$table->output($lang->upload_plugin);
+	$form->end();
+
 	if(!empty($plugins_list))
 	{
 		$a_plugins = $i_plugins = array();
 
-		foreach($plugins_list as $plugin_file)
+		if(!empty($plugins_list))
 		{
-			require_once MYBB_ROOT."inc/plugins/".$plugin_file;
-			$codename = str_replace(".php", "", $plugin_file);
-			$infofunc = $codename."_info";
-
-			if(!function_exists($infofunc))
+			foreach($plugins_list as $plugin_file)
 			{
-				continue;
-			}
-
-			$plugininfo = $infofunc();
-			$plugininfo['codename'] = $codename;
+				$codename = str_replace('.php', '', $plugin_file);
+				$plugininfo = read_json_file(MYBB_ROOT."inc/plugins/{$codename}/plugin.json", $err_msg, /*$show_errs = */false);
```

Comment:
I honestly would prefer if you would force the same structure for everything, including a certain folder structure inside the ZIP files. This a) would finally allow to script plugin installations (or ideally use composer for it) and b) simplify the whole plugin system and make things more maintainable. So if you already break the plugin system, take the chance and do it properly once instead of over and over again over the next couple releases.

Personally I would also implement a `PluginInterface` (along with either a trait or an abstract class for convenience) that defines `initialize`, `enable`, `disable`, `install`, `uninstall` etc methods and force the plugins to wrap their methods into a proper, ideally namespaced class. It's not really much more work for plugin authors to convert the currently global functions into a class, and on MyBB side you could get rid of all the `if(function_exists())` stuff and simply call it.

This class could in the end also hold the metadata and what not.